src/sender_queue/mod.rs

```diff
@@ -0,0 +1,366 @@
+    }
+
+    /// Removes all messages queued for the remote node from epochs upto `epoch`.
+    fn remove_earlier_messages(
+        &mut self,
+        sender_id: &D::NodeId,
+        epoch: <D::Message as Epoched>::Epoch,
+    ) {
+        let earlier_keys: Vec<_> = self
+            .outgoing_queue
+            .keys()
+            .cloned()
+            .filter(|(id, this_epoch)| {
+                id == sender_id
+                    && PartialOrd::partial_cmp(this_epoch, &epoch) == Some(Ordering::Less)
+            }).collect();
+        for key in earlier_keys {
+            self.outgoing_queue.remove(&key);
+        }
+    }
+
+    /// Processes an announcement of a new epoch update received from a remote node.
+    fn process_new_epoch(
+        &mut self,
+        sender_id: &D::NodeId,
+        epoch: <D::Message as Epoched>::Epoch,
+    ) -> Step<D> {
+        // Send any HB messages for the HB epoch.
+        let mut ready_messages = self
+            .outgoing_queue
+            .remove(&(sender_id.clone(), epoch))
+            .unwrap_or_else(|| vec![]);
+        for u in <D as SenderQueueFunc<D>>::spanning_epochs(epoch) {
+            // Send any DHB messages for the DHB era.
+            ready_messages.extend(
+                self.outgoing_queue
+                    .remove(&(sender_id.clone(), u))
+                    .unwrap_or_else(|| vec![]),
+            );
+        }
+        Step::from(ready_messages.into_iter().map(|msg| {
+            Target::Node(sender_id.clone()).message(Message {
+                epoch: msg.epoch(),
+                content: MessageContent::Algo(msg),
+            })
+        }))
+    }
+
+    /// Handles a Honey Badger algorithm message in a given epoch.
+    fn handle_message_content(
+        &mut self,
+        sender_id: &D::NodeId,
+        content: D::Message,
+    ) -> Result<Step<D>, D> {
+        let mut step = self.algo.handle_message(sender_id, content)?;
+        let mut sender_queue_step = self.update_epoch(&step);
+        self.defer_messages(&mut step);
+        sender_queue_step.extend(step.map(|output| output, Message::from));
+        Ok(sender_queue_step)
+    }
+
+    /// Updates the current Honey Badger epoch.
+    fn update_epoch(&mut self, step: &::Step<D>) -> Step<D> {
+        let mut updated = false;
+        // Look up `DynamicHoneyBadger` epoch updates and collect any added peers.
+        let (new_epoch, mut new_peers) = step.output.iter().fold(
+            (self.epoch, BTreeSet::new()),
+            |(epoch, mut new_peers), batch| {
+                let (max_epoch, new_peer) = self.algo.max_epoch_with_batch(epoch, batch);
+                if max_epoch != epoch {
+                    updated = true;
+                }
+                new_peer.map(|peer| new_peers.insert(peer));
+                (max_epoch, new_peers)
+            },
+        );
+        self.epoch = new_epoch;
+        self.peer_ids.append(&mut new_peers);
```

Comment:
Would it make sense to set their epoch to the next one right away?

Reply:
Possibly, I'm not sure. Is your idea that by doing this we would help the joining nodes to catch up faster?

Reply:
No, I'm just thinking that we already know what they new node is going to tell us: We already know that the batch's next epoch is the one we will receive in its `EpochStarted` message?
If not, we could set it to the default epoch. I'm just thinking it would make it easier if we didn't have the separate `peer_id`s collection, and just always had every peer as a key in the epoch map?